File: LAB_CN/LAB_CN/Python/hamming.py
def calculate_parity_bits(data):
    p1 = data[0] ^ data[1] ^ data[3]
    p2 = data[0] ^ data[2] ^ data[3]
    p3 = data[1] ^ data[2] ^ data[3]
    return p1, p2, p3


def add_parity_bits(data):
    p1, p2, p3 = calculate_parity_bits(data)
    data.insert(0, p1)
    data.insert(1, p2)
    data.insert(3, p3)


def detect_and_correct_error(received_data):
    error_position = 0
    p1 = received_data[0] ^ received_data[2] ^ received_data[4] ^ received_data[6]
    p2 = received_data[1] ^ received_data[2] ^ received_data[5] ^ received_data[6]
    p3 = received_data[3] ^ received_data[4] ^ received_data[5] ^ received_data[6]
    error_position = p3 * 4 + p2 * 2 + p1
    if error_position != 0:
        print("Error on the position:", error_position)
        received_data[error_position - 1] = 1 - received_data[error_position - 1]
        print("Correct message is:", ''.join(map(str, received_data)))
    else:
        print("Congratulations! There is no error.")
        print("Received data:", ''.join(map(str, received_data)))


# Main code
data = input("Enter 4 bits of data one by one (separated by spaces): ").split()
data = [int(bit) for bit in data]
add_parity_bits(data)
print("Encoded data is:", ''.join(map(str, data)))
received_data = input("Enter received data bits one by one (separated by spaces): ").split()
received_data = [int(bit) for bit in received_data]
detect_and_correct_error(received_data)

File: LAB_CN/LAB_CN/Java/Hamming Code/hamming.c
#include <stdio.h>

int main() {
    int data[8];
    printf("Enter the code to be sent one by one\n");
    scanf("%d %d %d %d",&data[3],&data[5],&data[6],&data[7]);
    data[1]=data[3]^data[5]^data[7];
    data[2]=data[3]^data[6]^data[7];
    data[4]=data[5]^data[6]^data[7];
    for(int i=1;i<=7;i++)
    {
        printf("%d",data[i]);
    }
    int datarec[8];
    printf("\nEnter a received code\n");
    for(int i=1;i<=7;i++)
    {
        scanf("%d",&datarec[i]);
        
    }
    int c1=datarec[1]^datarec[3]^datarec[5]^datarec[7];
    int c2=datarec[2]^datarec[3]^datarec[6]^datarec[7];
    int c3=datarec[4]^datarec[5]^datarec[6]^datarec[7];
    int c=c3*4+c2*2+c1;
    if(c==0)
    {
        printf("the error is not there");
        
    }
    else
    {
        printf("The error is found at %d ",c);
        if(datarec[c]==0)
        {
            datarec[c]=1;
        }
        else
        {
            datarec[c]=0;
        }
    }
    printf("the correct code is \n");
    for(int i=1;i<=7;i++)
    {
        printf("%d",datarec[i]);
    }
    
    
}  